
<h1 align=center><font size = 5>Segmenting and Clustering Crimes Against Females in New York City</font></h1>



This notebook was created by Amr Elrasad as part of Capstone project. 

## Introduction

In notebook, an analysis and exploration of the incident reports to New York policd department. The goal of the analysis is to analyize the crimes where victims were females and their locations. The goal is to provide analysis and clustering for places where crimes are more likely to be against women. 

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1. <a href="#item1">Download and Explore Dataset</a>
3. <a href="#item2">FOURSQUARE API</a>

2. <a href="#item2">Data Clustering</a>

</font>
</div>

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans
from sklearn.cluster import DBSCAN
from sklearn import preprocessing
#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


<a id='item1'></a>

In [2]:
# function to normalize dataframe before passing to clustering 
def normalize_df(df):
    x = df.values #returns a numpy array
    min_max_scaler = preprocessing.MinMaxScaler()
    x_scaled = min_max_scaler.fit_transform(x)
    df = pd.DataFrame(x_scaled)
    return df

## 1. Download and Explore Dataset

New York city has 77 precincts, where each incident report must include its unique code.

The national US government has open dataset for the all reported incidents between 2006 till 2017. This can be accessed from: https://catalog.data.gov/dataset/nypd-complaint-data-historic. The data is about 2GB and I only took subset o it for analysis. I choose to work on data in 2017.  


For your convenience, I downloaded the data and the subset to be analyzed is saved in 'data.csv'.

In [3]:
NYPD=pd.read_csv('data.csv')
NYPD.head()


CMPLNT_NUM CMPLNT_FR_DT CMPLNT_FR_TM CMPLNT_TO_DT CMPLNT_TO_TM  \
0   619128592   05/25/2017     11:00:00   05/25/2017     11:15:00   
1   699494668   05/25/2017     11:00:00   05/25/2017     12:00:00   
2   103321764   05/25/2017     11:00:00   05/25/2017     11:10:00   
3   137516053   05/25/2017     11:00:00   05/25/2017     11:05:00   
4   451320561   05/25/2017     11:00:00          NaN          NaN   

   ADDR_PCT_CD      RPT_DT  KY_CD                       OFNS_DESC  PD_CD  \
0           48  05/25/2017    351  CRIMINAL MISCHIEF & RELATED OF  259.0   
1           71  05/25/2017    106                  FELONY ASSAULT  109.0   
2          104  05/25/2017    352               CRIMINAL TRESPASS  205.0   
3          113  05/25/2017    341                   PETIT LARCENY  333.0   
4            5  05/25/2017    351  CRIMINAL MISCHIEF & RELATED OF  258.0   

                            PD_DESC CRM_ATPT_CPTD_CD   LAW_CAT_CD    BORO_NM  \
0  CRIMINAL MISCHIEF,UNCLASSIFIED 4        COMPLETED  MISDEMEANOR      BRONX   
1          ASSAULT 2,1,UNCLASSIFIED        COMPLETED       FELONY   BROOKLYN   
2              TRESPASS 2, CRIMINAL        COMPLETED  MISDEMEANOR     QUEENS   
3    LARCENY,PETIT FROM STORE-SHOPL        COMPLETED  MISDEMEANOR     QUEENS   
4    CRIMINAL MISCHIEF 4TH, GRAFFIT        COMPLETED  MISDEMEANOR  MANHATTAN   

  LOC_OF_OCCUR_DESC           PREM_TYP_DESC        JURIS_DESC  \
0            INSIDE     COMMERCIAL BUILDING  N.Y. POLICE DEPT   
1            INSIDE  RESIDENCE - APT. HOUSE  N.Y. POLICE DEPT   
2            INSIDE         RESIDENCE-HOUSE  N.Y. POLICE DEPT   
3            INSIDE        FOOD SUPERMARKET  N.Y. POLICE DEPT   
4          FRONT OF     COMMERCIAL BUILDING  N.Y. POLICE DEPT   

   JURISDICTION_CODE PARKS_NM HADEVELOPT  HOUSING_PSA  X_COORD_CD  Y_COORD_CD  \
0                0.0      NaN        NaN          NaN   1013619.0    247689.0   
1                0.0      NaN        NaN          NaN   1003717.0    182563.0   
2                0.0      NaN        NaN          NaN   1018935.0    205144.0   
3                0.0      NaN        NaN          NaN   1042058.0    186716.0   
4                0.0      NaN        NaN          NaN    984607.0    199399.0   

  SUSP_AGE_GROUP SUSP_RACE SUSP_SEX  TRANSIT_DISTRICT   Latitude  Longitude  \
0            NaN       NaN      NaN               NaN  40.846484 -73.893850   
1          25-44     BLACK        M               NaN  40.667757 -73.929829   
2          18-24     WHITE        F               NaN  40.729690 -73.874856   
3          25-44     BLACK        M               NaN  40.678989 -73.791585   
4            NaN       NaN      NaN               NaN  40.713989 -73.998714   

                         Lat_Lon               PATROL_BORO STATION_NAME  \
0  (40.846483584, -73.893850279)         PATROL BORO BRONX          NaN   
1  (40.667756807, -73.929828559)   PATROL BORO BKLYN SOUTH          NaN   
2  (40.729689793, -73.874855861)  PATROL BORO QUEENS NORTH          NaN   
3   (40.67898851, -73.791585115)  PATROL BORO QUEENS SOUTH          NaN   
4  (40.713989146, -73.998713673)     PATROL BORO MAN SOUTH          NaN   

  VIC_AGE_GROUP VIC_RACE VIC_SEX  
0       UNKNOWN  UNKNOWN       D  
1         18-24    BLACK       F  
2         45-64  UNKNOWN       F  
3       UNKNOWN  UNKNOWN       D  
4       UNKNOWN  UNKNOWN       E

Now, I am going to choose the columns to work on. Which are: Borough, crime category, precincts code, incident location, and victim sex. 

In [4]:
df_data=NYPD[['ADDR_PCT_CD','BORO_NM','Latitude','Longitude','LAW_CAT_CD','VIC_SEX']]
df_data.head()

ADDR_PCT_CD    BORO_NM   Latitude  Longitude   LAW_CAT_CD VIC_SEX
0           48      BRONX  40.846484 -73.893850  MISDEMEANOR       D
1           71   BROOKLYN  40.667757 -73.929829       FELONY       F
2          104     QUEENS  40.729690 -73.874856  MISDEMEANOR       F
3          113     QUEENS  40.678989 -73.791585  MISDEMEANOR       D
4            5  MANHATTAN  40.713989 -73.998714  MISDEMEANOR       E

In [5]:
len(df_data)

168568

It seems that there were 168568 incidents in the first 5 months of 2017!!

Now let us consider only incidents where the victim was female.

In [6]:
df_data=df_data[df_data['VIC_SEX']=='F']
df_data.drop(labels='VIC_SEX',axis=1,inplace=True)
df_data=df_data.reset_index(drop=True)
len(df_data)

66386

There were 66386 incidents against women.
Let us view the data that we will work on it

In [7]:
df_data.head()

ADDR_PCT_CD    BORO_NM   Latitude  Longitude   LAW_CAT_CD
0           71   BROOKLYN  40.667757 -73.929829       FELONY
1          104     QUEENS  40.729690 -73.874856  MISDEMEANOR
2           48      BRONX  40.855168 -73.887904  MISDEMEANOR
3           45      BRONX  40.842202 -73.849757  MISDEMEANOR
4           24  MANHATTAN  40.794953 -73.971437       FELONY

Data types:

In [8]:
df_data.dtypes

ADDR_PCT_CD      int64
BORO_NM         object
Latitude       float64
Longitude      float64
LAW_CAT_CD      object
dtype: object

Now let us drop the missing data

In [9]:
df_data.BORO_NM.replace(np.nan, '', regex=True,inplace=True)
df_data=df_data[df_data['BORO_NM']!=''].reset_index(drop=True)
df_data=df_data[df_data['ADDR_PCT_CD']!=np.nan].reset_index(drop=True)
print(len(df_data))


66365


It is time for one-hot encoding for the offence category. Also, we will group data by precincts for better viewing   

In [10]:
df_data_onehot= pd.get_dummies(df_data[['LAW_CAT_CD']], prefix="", prefix_sep="")
df_data=pd.concat([df_data,df_data_onehot],axis=1).drop(['LAW_CAT_CD'],axis=1)
df_data=df_data.sort_values(by=['ADDR_PCT_CD']).reset_index(drop=True)
df_data.head()

ADDR_PCT_CD    BORO_NM   Latitude  Longitude  FELONY  MISDEMEANOR  \
0            1  MANHATTAN  40.710229 -74.007746       0            0   
1            1  MANHATTAN  40.720464 -74.006852       0            1   
2            1  MANHATTAN  40.715529 -74.009240       0            1   
3            1  MANHATTAN  40.722855 -74.003375       1            0   
4            1  MANHATTAN  40.705024 -74.012978       0            1   

   VIOLATION  
0          1  
1          0  
2          0  
3          0  
4          0

It is better to group data by precincts and borough and count total number of crimes for each category and also calculate average location to place it on map.  

In [11]:
df_grouped1=df_data.groupby(by=['ADDR_PCT_CD','BORO_NM']).count().reset_index()
df_grouped2=df_data.groupby(by=['ADDR_PCT_CD','BORO_NM']).mean().reset_index()
df_grouped3=df_data.groupby(by=['ADDR_PCT_CD','BORO_NM']).sum().reset_index()


Now  lust us construct the data frame and the one that will hold the clusters labels

I also added score metric column to reflect the number of crimes and the types of them. For each Felony, I give score of 5, Misdemeanor score of 3, and Violation score of 2. 

In [12]:
df_grouped1['Latitude']=df_grouped2['Latitude']
df_grouped1['Longitude']=df_grouped2['Longitude']
df_grouped1['FELONY']=df_grouped3['FELONY']
df_grouped1['MISDEMEANOR']=df_grouped3['MISDEMEANOR']
df_grouped1['VIOLATION']=df_grouped3['VIOLATION']
df_grouped=df_grouped1
df_grouped['Score']=df_grouped['FELONY']*5+df_grouped['MISDEMEANOR']*3+df_grouped['VIOLATION']*2
del df_grouped1, df_grouped2,df_grouped3

In [13]:
df_grouped

ADDR_PCT_CD        BORO_NM   Latitude  Longitude  FELONY  MISDEMEANOR  \
0             1      MANHATTAN  40.714436 -74.008120   193.0        284.0   
1             5      MANHATTAN  40.718126 -73.995498   160.0        254.0   
2             6      MANHATTAN  40.733790 -74.001023   233.0        205.0   
3             7      MANHATTAN  40.716364 -73.984518   163.0        294.0   
4             9      MANHATTAN  40.726165 -73.983405   273.0        369.0   
5             9         QUEENS  40.726554 -73.987828     1.0          0.0   
6            10      MANHATTAN  40.747869 -74.000401   177.0        224.0   
7            13      MANHATTAN  40.738675 -73.985680   281.0        335.0   
8            14      MANHATTAN  40.752727 -73.988398   405.0        381.0   
9            17      MANHATTAN  40.752853 -73.971203   178.0        156.0   
10           18      MANHATTAN  40.763205 -73.985063   302.0        357.0   
11           19      MANHATTAN  40.772930 -73.956127   370.0        312.0   
12           20      MANHATTAN  40.778830 -73.980717   197.0        204.0   
13           22      MANHATTAN  40.781662 -73.966647    13.0          9.0   
14           23      MANHATTAN  40.791940 -73.945005   201.0        480.0   
15           24      MANHATTAN  40.795593 -73.968273   250.0        251.0   
16           25      MANHATTAN  40.802589 -73.937270   227.0        401.0   
17           26      MANHATTAN  40.812321 -73.957500   118.0        229.0   
18           28      MANHATTAN  40.804835 -73.950847   193.0        344.0   
19           30      MANHATTAN  40.825716 -73.948744   152.0        224.0   
20           32      MANHATTAN  40.819637 -73.941336   205.0        525.0   
21           33      MANHATTAN  40.839890 -73.939988   143.0        244.0   
22           34      MANHATTAN  40.858572 -73.928164   303.0        377.0   
23           40          BRONX  40.813599 -73.917341   419.0        899.0   
24           40      MANHATTAN  40.810398 -73.924895     1.0          0.0   
25           41          BRONX  40.818131 -73.893347   183.0        463.0   
26           42          BRONX  40.830880 -73.900819   271.0        625.0   
27           43          BRONX  40.826629 -73.864702   379.0       1010.0   
28           44          BRONX  40.834812 -73.918819   390.0        934.0   
29           45          BRONX  40.843702 -73.828042   189.0        479.0   
30           46          BRONX  40.852742 -73.907914   364.0        752.0   
31           47          BRONX  40.885309 -73.852984   513.0        891.0   
32           48          BRONX  40.849150 -73.888063   331.0        622.0   
33           49          BRONX  40.857817 -73.859207   258.0        441.0   
34           50          BRONX  40.882525 -73.904187   214.0        334.0   
35           52          BRONX  40.870076 -73.889549   325.0        798.0   
36           60       BROOKLYN  40.579061 -73.982817   209.0        428.0   
37           61       BROOKLYN  40.595714 -73.951280   254.0        360.0   
38           62       BROOKLYN  40.607097 -73.993810   226.0        341.0   
39           63       BROOKLYN  40.622848 -73.927110   175.0        347.0   
40           66       BROOKLYN  40.633077 -73.984669   181.0        271.0   
41           67       BROOKLYN  40.650404 -73.934986   473.0        662.0   
42           68       BROOKLYN  40.626373 -74.023181   187.0        265.0   
43           69       BROOKLYN  40.639674 -73.899548   179.0        263.0   
44           70       BROOKLYN  40.637004 -73.959707   323.0        481.0   
45           71       BROOKLYN  40.663246 -73.948029   268.0        428.0   
46           72       BROOKLYN  40.649712 -74.004699   191.0        317.0   
47           73       BROOKLYN  40.668572 -73.910828   391.0        738.0   
48           75       BROOKLYN  40.668280 -73.882312   704.0       1138.0   
49           76       BROOKLYN  40.679942 -74.000346   103.0        145.0   
50           77       BROOKLYN  40.673646 -73.940635   276.0        473.0   


Now, it is time to clean data from some repetitions at indcies: 5, 24, 67, and 74.

In [14]:
df_grouped=df_grouped.drop([5,24,67,74]).reset_index(drop=True)
df_grouped

ADDR_PCT_CD        BORO_NM   Latitude  Longitude  FELONY  MISDEMEANOR  \
0             1      MANHATTAN  40.714436 -74.008120   193.0        284.0   
1             5      MANHATTAN  40.718126 -73.995498   160.0        254.0   
2             6      MANHATTAN  40.733790 -74.001023   233.0        205.0   
3             7      MANHATTAN  40.716364 -73.984518   163.0        294.0   
4             9      MANHATTAN  40.726165 -73.983405   273.0        369.0   
5            10      MANHATTAN  40.747869 -74.000401   177.0        224.0   
6            13      MANHATTAN  40.738675 -73.985680   281.0        335.0   
7            14      MANHATTAN  40.752727 -73.988398   405.0        381.0   
8            17      MANHATTAN  40.752853 -73.971203   178.0        156.0   
9            18      MANHATTAN  40.763205 -73.985063   302.0        357.0   
10           19      MANHATTAN  40.772930 -73.956127   370.0        312.0   
11           20      MANHATTAN  40.778830 -73.980717   197.0        204.0   
12           22      MANHATTAN  40.781662 -73.966647    13.0          9.0   
13           23      MANHATTAN  40.791940 -73.945005   201.0        480.0   
14           24      MANHATTAN  40.795593 -73.968273   250.0        251.0   
15           25      MANHATTAN  40.802589 -73.937270   227.0        401.0   
16           26      MANHATTAN  40.812321 -73.957500   118.0        229.0   
17           28      MANHATTAN  40.804835 -73.950847   193.0        344.0   
18           30      MANHATTAN  40.825716 -73.948744   152.0        224.0   
19           32      MANHATTAN  40.819637 -73.941336   205.0        525.0   
20           33      MANHATTAN  40.839890 -73.939988   143.0        244.0   
21           34      MANHATTAN  40.858572 -73.928164   303.0        377.0   
22           40          BRONX  40.813599 -73.917341   419.0        899.0   
23           41          BRONX  40.818131 -73.893347   183.0        463.0   
24           42          BRONX  40.830880 -73.900819   271.0        625.0   
25           43          BRONX  40.826629 -73.864702   379.0       1010.0   
26           44          BRONX  40.834812 -73.918819   390.0        934.0   
27           45          BRONX  40.843702 -73.828042   189.0        479.0   
28           46          BRONX  40.852742 -73.907914   364.0        752.0   
29           47          BRONX  40.885309 -73.852984   513.0        891.0   
30           48          BRONX  40.849150 -73.888063   331.0        622.0   
31           49          BRONX  40.857817 -73.859207   258.0        441.0   
32           50          BRONX  40.882525 -73.904187   214.0        334.0   
33           52          BRONX  40.870076 -73.889549   325.0        798.0   
34           60       BROOKLYN  40.579061 -73.982817   209.0        428.0   
35           61       BROOKLYN  40.595714 -73.951280   254.0        360.0   
36           62       BROOKLYN  40.607097 -73.993810   226.0        341.0   
37           63       BROOKLYN  40.622848 -73.927110   175.0        347.0   
38           66       BROOKLYN  40.633077 -73.984669   181.0        271.0   
39           67       BROOKLYN  40.650404 -73.934986   473.0        662.0   
40           68       BROOKLYN  40.626373 -74.023181   187.0        265.0   
41           69       BROOKLYN  40.639674 -73.899548   179.0        263.0   
42           70       BROOKLYN  40.637004 -73.959707   323.0        481.0   
43           71       BROOKLYN  40.663246 -73.948029   268.0        428.0   
44           72       BROOKLYN  40.649712 -74.004699   191.0        317.0   
45           73       BROOKLYN  40.668572 -73.910828   391.0        738.0   
46           75       BROOKLYN  40.668280 -73.882312   704.0       1138.0   
47           76       BROOKLYN  40.679942 -74.000346   103.0        145.0   
48           77       BROOKLYN  40.673646 -73.940635   276.0        473.0   
49           78       BROOKLYN  40.674806 -73.978710   178.0        241.0   
50           79       BROOKLYN  40.689410 -73.947560   258.0        479.0   


The clustered grouped data frame is the one that will hold the data with clusters labels 

In [15]:
df_grouped_clusters=df_grouped.copy()
df_grouped_clusters.insert(0,'Cluster Labels',np.nan)
df_grouped_clusters.head()

Cluster Labels  ADDR_PCT_CD    BORO_NM   Latitude  Longitude  FELONY  \
0             NaN            1  MANHATTAN  40.714436 -74.008120   193.0   
1             NaN            5  MANHATTAN  40.718126 -73.995498   160.0   
2             NaN            6  MANHATTAN  40.733790 -74.001023   233.0   
3             NaN            7  MANHATTAN  40.716364 -73.984518   163.0   
4             NaN            9  MANHATTAN  40.726165 -73.983405   273.0   

   MISDEMEANOR  VIOLATION   Score  
0        284.0      113.0  2043.0  
1        254.0       87.0  1736.0  
2        205.0       78.0  1936.0  
3        294.0      132.0  1961.0  
4        369.0      102.0  2676.0

# 2. FOURSQUAR API

I will use foursquare API to get correct coordinates of each precinct. I have created a list with all precincts on my foursquare account. I am going to get and parse this list to extract the coordinates and update both grouped and clustered grouped data frames. 

## 2.1 Credentials

In [16]:
CLIENT_ID = 'TEESM4UGUK3OBYJ5BCHFZAM5SCUCDQWDRIQH4IZGQEW13GAL' # your Foursquare ID
CLIENT_SECRET = 'ERL1CINJIPLCEL3RPMALZYFOHJOCCV1PZYV2YXETTM5DZHS4' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 2000

In [17]:
list_id = '5c8d3fdb1acf11002c667a3a' # user ID with most agree counts and complete profile
url = 'https://api.foursquare.com/v2/lists/{}?client_id={}&client_secret={}&v={}'.format(list_id, CLIENT_ID, CLIENT_SECRET, VERSION) # define URL
# send GET request
results = requests.get(url).json()

In [18]:
for i in range(0,77):
    lng=results['response']['list']['listItems']['items'][i]['venue']['location']['labeledLatLngs'][0]['lng']
    lat=results['response']['list']['listItems']['items'][i]['venue']['location']['labeledLatLngs'][0]['lat']
    df_grouped.loc[i,'Latitude']=lat
    df_grouped.loc[i,'Longitude']=lng
    df_grouped_clusters.loc[i,'Latitude']=lat
    df_grouped_clusters.loc[i,'Longitude']=lng


In [19]:
df_grouped

ADDR_PCT_CD        BORO_NM   Latitude  Longitude  FELONY  MISDEMEANOR  \
0             1      MANHATTAN  40.720384 -74.006939   193.0        284.0   
1             5      MANHATTAN  40.716097 -73.997252   160.0        254.0   
2             6      MANHATTAN  40.733985 -74.005457   233.0        205.0   
3             7      MANHATTAN  40.716392 -73.983726   163.0        294.0   
4             9      MANHATTAN  40.726359 -73.988002   273.0        369.0   
5            10      MANHATTAN  40.742876 -73.998551   177.0        224.0   
6            13      MANHATTAN  40.736980 -73.982771   281.0        335.0   
7            14      MANHATTAN  40.753830 -73.995050   405.0        381.0   
8            17      MANHATTAN  40.756657 -73.970653   178.0        156.0   
9            18      MANHATTAN  40.765130 -73.985013   302.0        357.0   
10           19      MANHATTAN  40.767065 -73.963994   370.0        312.0   
11           20      MANHATTAN  40.784382 -73.975109   197.0        204.0   
12           22      MANHATTAN  40.784413 -73.967138    13.0          9.0   
13           23      MANHATTAN  40.789068 -73.946818   201.0        480.0   
14           24      MANHATTAN  40.794691 -73.967040   250.0        251.0   
15           25      MANHATTAN  40.800868 -73.941000   227.0        401.0   
16           26      MANHATTAN  40.814682 -73.956520   118.0        229.0   
17           28      MANHATTAN  40.808560 -73.952465   193.0        344.0   
18           30      MANHATTAN  40.828701 -73.943991   152.0        224.0   
19           32      MANHATTAN  40.815997 -73.945171   205.0        525.0   
20           33      MANHATTAN  40.840992 -73.935974   143.0        244.0   
21           34      MANHATTAN  40.851259 -73.935131   303.0        377.0   
22           40          BRONX  40.810369 -73.925339   419.0        899.0   
23           41          BRONX  40.816429 -73.895458   183.0        463.0   
24           42          BRONX  40.822231 -73.911447   271.0        625.0   
25           43          BRONX  40.823195 -73.870354   379.0       1010.0   
26           44          BRONX  40.837659 -73.920010   390.0        934.0   
27           45          BRONX  40.830830 -73.827246   189.0        479.0   
28           46          BRONX  40.854114 -73.900178   364.0        752.0   
29           47          BRONX  40.887270 -73.847293   513.0        891.0   
30           48          BRONX  40.844134 -73.900270   331.0        622.0   
31           49          BRONX  40.855957 -73.844251   258.0        441.0   
32           50          BRONX  40.883651 -73.902314   214.0        334.0   
33           52          BRONX  40.868664 -73.879806   325.0        798.0   
34           60       BROOKLYN  40.576508 -73.976517   209.0        428.0   
35           61       BROOKLYN  40.594119 -73.960828   254.0        360.0   
36           62       BROOKLYN  40.602419 -74.002898   226.0        341.0   
37           63       BROOKLYN  40.628032 -73.941419   175.0        347.0   
38           66       BROOKLYN  40.625653 -73.991293   181.0        271.0   
39           67       BROOKLYN  40.648796 -73.950307   473.0        662.0   
40           68       BROOKLYN  40.639290 -74.022785   187.0        265.0   
41           69       BROOKLYN  40.648685 -73.905188   179.0        263.0   
42           70       BROOKLYN  40.630426 -73.973779   323.0        481.0   
43           71       BROOKLYN  40.664657 -73.948237   268.0        428.0   
44           72       BROOKLYN  40.658890 -74.001181   191.0        317.0   
45           73       BROOKLYN  40.670889 -73.913899   391.0        738.0   
46           75       BROOKLYN  40.671193 -73.881505   704.0       1138.0   
47           76       BROOKLYN  40.683614 -74.000185   103.0        145.0   
48           77       BROOKLYN  40.675023 -73.930195   276.0        473.0   
49           78       BROOKLYN  40.680827 -73.974471   178.0        241.0   
50           79       BROOKLYN  40.688798 -73.944957   258.0        479.0   


In [20]:
df_grouped_clusters

Cluster Labels  ADDR_PCT_CD        BORO_NM   Latitude  Longitude  FELONY  \
0              NaN            1      MANHATTAN  40.720384 -74.006939   193.0   
1              NaN            5      MANHATTAN  40.716097 -73.997252   160.0   
2              NaN            6      MANHATTAN  40.733985 -74.005457   233.0   
3              NaN            7      MANHATTAN  40.716392 -73.983726   163.0   
4              NaN            9      MANHATTAN  40.726359 -73.988002   273.0   
5              NaN           10      MANHATTAN  40.742876 -73.998551   177.0   
6              NaN           13      MANHATTAN  40.736980 -73.982771   281.0   
7              NaN           14      MANHATTAN  40.753830 -73.995050   405.0   
8              NaN           17      MANHATTAN  40.756657 -73.970653   178.0   
9              NaN           18      MANHATTAN  40.765130 -73.985013   302.0   
10             NaN           19      MANHATTAN  40.767065 -73.963994   370.0   
11             NaN           20      MANHATTAN  40.784382 -73.975109   197.0   
12             NaN           22      MANHATTAN  40.784413 -73.967138    13.0   
13             NaN           23      MANHATTAN  40.789068 -73.946818   201.0   
14             NaN           24      MANHATTAN  40.794691 -73.967040   250.0   
15             NaN           25      MANHATTAN  40.800868 -73.941000   227.0   
16             NaN           26      MANHATTAN  40.814682 -73.956520   118.0   
17             NaN           28      MANHATTAN  40.808560 -73.952465   193.0   
18             NaN           30      MANHATTAN  40.828701 -73.943991   152.0   
19             NaN           32      MANHATTAN  40.815997 -73.945171   205.0   
20             NaN           33      MANHATTAN  40.840992 -73.935974   143.0   
21             NaN           34      MANHATTAN  40.851259 -73.935131   303.0   
22             NaN           40          BRONX  40.810369 -73.925339   419.0   
23             NaN           41          BRONX  40.816429 -73.895458   183.0   
24             NaN           42          BRONX  40.822231 -73.911447   271.0   
25             NaN           43          BRONX  40.823195 -73.870354   379.0   
26             NaN           44          BRONX  40.837659 -73.920010   390.0   
27             NaN           45          BRONX  40.830830 -73.827246   189.0   
28             NaN           46          BRONX  40.854114 -73.900178   364.0   
29             NaN           47          BRONX  40.887270 -73.847293   513.0   
30             NaN           48          BRONX  40.844134 -73.900270   331.0   
31             NaN           49          BRONX  40.855957 -73.844251   258.0   
32             NaN           50          BRONX  40.883651 -73.902314   214.0   
33             NaN           52          BRONX  40.868664 -73.879806   325.0   
34             NaN           60       BROOKLYN  40.576508 -73.976517   209.0   
35             NaN           61       BROOKLYN  40.594119 -73.960828   254.0   
36             NaN           62       BROOKLYN  40.602419 -74.002898   226.0   
37             NaN           63       BROOKLYN  40.628032 -73.941419   175.0   
38             NaN           66       BROOKLYN  40.625653 -73.991293   181.0   
39             NaN           67       BROOKLYN  40.648796 -73.950307   473.0   
40             NaN           68       BROOKLYN  40.639290 -74.022785   187.0   
41             NaN           69       BROOKLYN  40.648685 -73.905188   179.0   
42             NaN           70       BROOKLYN  40.630426 -73.973779   323.0   
43             NaN           71       BROOKLYN  40.664657 -73.948237   268.0   
44             NaN           72       BROOKLYN  40.658890 -74.001181   191.0   
45             NaN           73       BROOKLYN  40.670889 -73.913899   391.0   
46             NaN           75       BROOKLYN  40.671193 -73.881505   704.0   
47             NaN           76       BROOKLYN  40.683614 -74.000185   103.0   
48             NaN           77       BROOKLYN  40.675023 -73.930195   276.0   
49  

# 3. Data Clustering

## 3.1 Map

This function will produce folium map.

In [21]:
def create_folium_map():
    # create map
    map_clusters = folium.Map(location=[40.7308619, -73.9871558], zoom_start=11)

    # set color scheme for the clusters
    x = np.arange(kclusters)
    ys = [i + x + (i*x)**2 for i in range(kclusters)]
    colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
    rainbow = [colors.rgb2hex(i) for i in colors_array]

    # add markers to the map
    markers_colors = []
    for lat, lon, poi, boro, cluster,score in zip(df_grouped_clusters['Latitude'], df_grouped_clusters['Longitude'], df_grouped_clusters['ADDR_PCT_CD'],df_grouped_clusters['BORO_NM'], df_grouped_clusters['Cluster Labels'],df_grouped_clusters['Score']):
        label = folium.Popup('precinct No.  '+str(poi) +', '+boro+ ' Cluster ' + str(cluster)+ ', score= '+ str(score), parse_html=True)
        folium.CircleMarker(
            [lat, lon],
            radius=5,
            popup=label,
            color=rainbow[cluster-1],
            fill=True,
            fill_color=rainbow[cluster-1],
            fill_opacity=0.7).add_to(map_clusters)
    return map_clusters

## 3.2 K-means

K-means clustering is used to cluster data based on the 'df_grouped' data frame. 

### 3.2.1 Clustering based on location and number of each crime category 

In [22]:
# set number of clusters
kclusters = 3

# run k-means clustering
# normalizing values first
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(normalize_df(df_grouped[['Latitude','Longitude','FELONY','MISDEMEANOR','VIOLATION']]))

# add clustering labels
df_grouped_clusters['Cluster Labels']= kmeans.labels_
# view clusters data frame
df_grouped_clusters # check the last columns!

Cluster Labels  ADDR_PCT_CD        BORO_NM   Latitude  Longitude  FELONY  \
0                1            1      MANHATTAN  40.720384 -74.006939   193.0   
1                1            5      MANHATTAN  40.716097 -73.997252   160.0   
2                1            6      MANHATTAN  40.733985 -74.005457   233.0   
3                1            7      MANHATTAN  40.716392 -73.983726   163.0   
4                1            9      MANHATTAN  40.726359 -73.988002   273.0   
5                1           10      MANHATTAN  40.742876 -73.998551   177.0   
6                1           13      MANHATTAN  40.736980 -73.982771   281.0   
7                1           14      MANHATTAN  40.753830 -73.995050   405.0   
8                1           17      MANHATTAN  40.756657 -73.970653   178.0   
9                1           18      MANHATTAN  40.765130 -73.985013   302.0   
10               1           19      MANHATTAN  40.767065 -73.963994   370.0   
11               1           20      MANHATTAN  40.784382 -73.975109   197.0   
12               1           22      MANHATTAN  40.784413 -73.967138    13.0   
13               1           23      MANHATTAN  40.789068 -73.946818   201.0   
14               1           24      MANHATTAN  40.794691 -73.967040   250.0   
15               1           25      MANHATTAN  40.800868 -73.941000   227.0   
16               1           26      MANHATTAN  40.814682 -73.956520   118.0   
17               1           28      MANHATTAN  40.808560 -73.952465   193.0   
18               1           30      MANHATTAN  40.828701 -73.943991   152.0   
19               1           32      MANHATTAN  40.815997 -73.945171   205.0   
20               1           33      MANHATTAN  40.840992 -73.935974   143.0   
21               1           34      MANHATTAN  40.851259 -73.935131   303.0   
22               2           40          BRONX  40.810369 -73.925339   419.0   
23               1           41          BRONX  40.816429 -73.895458   183.0   
24               2           42          BRONX  40.822231 -73.911447   271.0   
25               2           43          BRONX  40.823195 -73.870354   379.0   
26               2           44          BRONX  40.837659 -73.920010   390.0   
27               2           45          BRONX  40.830830 -73.827246   189.0   
28               2           46          BRONX  40.854114 -73.900178   364.0   
29               2           47          BRONX  40.887270 -73.847293   513.0   
30               2           48          BRONX  40.844134 -73.900270   331.0   
31               1           49          BRONX  40.855957 -73.844251   258.0   
32               1           50          BRONX  40.883651 -73.902314   214.0   
33               2           52          BRONX  40.868664 -73.879806   325.0   
34               0           60       BROOKLYN  40.576508 -73.976517   209.0   
35               0           61       BROOKLYN  40.594119 -73.960828   254.0   
36               0           62       BROOKLYN  40.602419 -74.002898   226.0   
37               0           63       BROOKLYN  40.628032 -73.941419   175.0   
38               0           66       BROOKLYN  40.625653 -73.991293   181.0   
39               2           67       BROOKLYN  40.648796 -73.950307   473.0   
40               0           68       BROOKLYN  40.639290 -74.022785   187.0   
41               0           69       BROOKLYN  40.648685 -73.905188   179.0   
42               0           70       BROOKLYN  40.630426 -73.973779   323.0   
43               0           71       BROOKLYN  40.664657 -73.948237   268.0   
44               0           72       BROOKLYN  40.658890 -74.001181   191.0   
45               2           73       BROOKLYN  40.670889 -73.913899   391.0   
46               2           75       BROOKLYN  40.671193 -73.881505   704.0   
47               1           76       BROOKLYN  40.683614 -74.000185   103.0   
48               0           77       BROOKLYN  40.675023 -73.930195   276.0   
49  

Now let us visualize clustered data on map

In [23]:
create_folium_map()

### 3.2.2 Clustering based on location and crime score 

In [24]:
# set number of clusters
kclusters = 3

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(normalize_df(df_grouped[['Latitude','Longitude','Score']]))

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:30] 
# add clustering labels
df_grouped_clusters['Cluster Labels']= kmeans.labels_

df_grouped_clusters 

Cluster Labels  ADDR_PCT_CD        BORO_NM   Latitude  Longitude  FELONY  \
0                1            1      MANHATTAN  40.720384 -74.006939   193.0   
1                1            5      MANHATTAN  40.716097 -73.997252   160.0   
2                1            6      MANHATTAN  40.733985 -74.005457   233.0   
3                1            7      MANHATTAN  40.716392 -73.983726   163.0   
4                1            9      MANHATTAN  40.726359 -73.988002   273.0   
5                1           10      MANHATTAN  40.742876 -73.998551   177.0   
6                1           13      MANHATTAN  40.736980 -73.982771   281.0   
7                2           14      MANHATTAN  40.753830 -73.995050   405.0   
8                1           17      MANHATTAN  40.756657 -73.970653   178.0   
9                2           18      MANHATTAN  40.765130 -73.985013   302.0   
10               2           19      MANHATTAN  40.767065 -73.963994   370.0   
11               2           20      MANHATTAN  40.784382 -73.975109   197.0   
12               1           22      MANHATTAN  40.784413 -73.967138    13.0   
13               2           23      MANHATTAN  40.789068 -73.946818   201.0   
14               2           24      MANHATTAN  40.794691 -73.967040   250.0   
15               2           25      MANHATTAN  40.800868 -73.941000   227.0   
16               2           26      MANHATTAN  40.814682 -73.956520   118.0   
17               2           28      MANHATTAN  40.808560 -73.952465   193.0   
18               2           30      MANHATTAN  40.828701 -73.943991   152.0   
19               2           32      MANHATTAN  40.815997 -73.945171   205.0   
20               2           33      MANHATTAN  40.840992 -73.935974   143.0   
21               2           34      MANHATTAN  40.851259 -73.935131   303.0   
22               2           40          BRONX  40.810369 -73.925339   419.0   
23               2           41          BRONX  40.816429 -73.895458   183.0   
24               2           42          BRONX  40.822231 -73.911447   271.0   
25               2           43          BRONX  40.823195 -73.870354   379.0   
26               2           44          BRONX  40.837659 -73.920010   390.0   
27               2           45          BRONX  40.830830 -73.827246   189.0   
28               2           46          BRONX  40.854114 -73.900178   364.0   
29               2           47          BRONX  40.887270 -73.847293   513.0   
30               2           48          BRONX  40.844134 -73.900270   331.0   
31               2           49          BRONX  40.855957 -73.844251   258.0   
32               2           50          BRONX  40.883651 -73.902314   214.0   
33               2           52          BRONX  40.868664 -73.879806   325.0   
34               1           60       BROOKLYN  40.576508 -73.976517   209.0   
35               1           61       BROOKLYN  40.594119 -73.960828   254.0   
36               1           62       BROOKLYN  40.602419 -74.002898   226.0   
37               1           63       BROOKLYN  40.628032 -73.941419   175.0   
38               1           66       BROOKLYN  40.625653 -73.991293   181.0   
39               0           67       BROOKLYN  40.648796 -73.950307   473.0   
40               1           68       BROOKLYN  40.639290 -74.022785   187.0   
41               1           69       BROOKLYN  40.648685 -73.905188   179.0   
42               1           70       BROOKLYN  40.630426 -73.973779   323.0   
43               1           71       BROOKLYN  40.664657 -73.948237   268.0   
44               1           72       BROOKLYN  40.658890 -74.001181   191.0   
45               0           73       BROOKLYN  40.670889 -73.913899   391.0   
46               0           75       BROOKLYN  40.671193 -73.881505   704.0   
47               1           76       BROOKLYN  40.683614 -74.000185   103.0   
48               1           77       BROOKLYN  40.675023 -73.930195   276.0   
49  

In [25]:
create_folium_map()

### 3.2.2 Clustering based on  crime score 

In [26]:
# set number of clusters
kclusters = 3

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(normalize_df(df_grouped[['Score']]))

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:30] 
# add clustering labels
df_grouped_clusters['Cluster Labels']= kmeans.labels_

df_grouped_clusters 

Cluster Labels  ADDR_PCT_CD        BORO_NM   Latitude  Longitude  FELONY  \
0                0            1      MANHATTAN  40.720384 -74.006939   193.0   
1                0            5      MANHATTAN  40.716097 -73.997252   160.0   
2                0            6      MANHATTAN  40.733985 -74.005457   233.0   
3                0            7      MANHATTAN  40.716392 -73.983726   163.0   
4                2            9      MANHATTAN  40.726359 -73.988002   273.0   
5                0           10      MANHATTAN  40.742876 -73.998551   177.0   
6                2           13      MANHATTAN  40.736980 -73.982771   281.0   
7                2           14      MANHATTAN  40.753830 -73.995050   405.0   
8                0           17      MANHATTAN  40.756657 -73.970653   178.0   
9                2           18      MANHATTAN  40.765130 -73.985013   302.0   
10               2           19      MANHATTAN  40.767065 -73.963994   370.0   
11               0           20      MANHATTAN  40.784382 -73.975109   197.0   
12               0           22      MANHATTAN  40.784413 -73.967138    13.0   
13               2           23      MANHATTAN  40.789068 -73.946818   201.0   
14               0           24      MANHATTAN  40.794691 -73.967040   250.0   
15               2           25      MANHATTAN  40.800868 -73.941000   227.0   
16               0           26      MANHATTAN  40.814682 -73.956520   118.0   
17               0           28      MANHATTAN  40.808560 -73.952465   193.0   
18               0           30      MANHATTAN  40.828701 -73.943991   152.0   
19               2           32      MANHATTAN  40.815997 -73.945171   205.0   
20               0           33      MANHATTAN  40.840992 -73.935974   143.0   
21               2           34      MANHATTAN  40.851259 -73.935131   303.0   
22               1           40          BRONX  40.810369 -73.925339   419.0   
23               2           41          BRONX  40.816429 -73.895458   183.0   
24               2           42          BRONX  40.822231 -73.911447   271.0   
25               1           43          BRONX  40.823195 -73.870354   379.0   
26               1           44          BRONX  40.837659 -73.920010   390.0   
27               2           45          BRONX  40.830830 -73.827246   189.0   
28               1           46          BRONX  40.854114 -73.900178   364.0   
29               1           47          BRONX  40.887270 -73.847293   513.0   
30               2           48          BRONX  40.844134 -73.900270   331.0   
31               2           49          BRONX  40.855957 -73.844251   258.0   
32               0           50          BRONX  40.883651 -73.902314   214.0   
33               1           52          BRONX  40.868664 -73.879806   325.0   
34               2           60       BROOKLYN  40.576508 -73.976517   209.0   
35               2           61       BROOKLYN  40.594119 -73.960828   254.0   
36               2           62       BROOKLYN  40.602419 -74.002898   226.0   
37               0           63       BROOKLYN  40.628032 -73.941419   175.0   
38               0           66       BROOKLYN  40.625653 -73.991293   181.0   
39               1           67       BROOKLYN  40.648796 -73.950307   473.0   
40               0           68       BROOKLYN  40.639290 -74.022785   187.0   
41               0           69       BROOKLYN  40.648685 -73.905188   179.0   
42               2           70       BROOKLYN  40.630426 -73.973779   323.0   
43               2           71       BROOKLYN  40.664657 -73.948237   268.0   
44               0           72       BROOKLYN  40.658890 -74.001181   191.0   
45               1           73       BROOKLYN  40.670889 -73.913899   391.0   
46               1           75       BROOKLYN  40.671193 -73.881505   704.0   
47               0           76       BROOKLYN  40.683614 -74.000185   103.0   
48               2           77       BROOKLYN  40.675023 -73.930195   276.0   
49  

In [27]:
create_folium_map()

## 3.2 DBSCAN Clustering

Another way to do clustering by using DBSCAN clustering

In [28]:
# run DBSCAN clustering
Dbscan = DBSCAN(eps=0.01,min_samples=5).fit(normalize_df(df_grouped[['Score']])) 
# add clustering labels
df_grouped_clusters['Cluster Labels']= Dbscan.labels_
df_grouped_clusters 

Cluster Labels  ADDR_PCT_CD        BORO_NM   Latitude  Longitude  FELONY  \
0                0            1      MANHATTAN  40.720384 -74.006939   193.0   
1               -1            5      MANHATTAN  40.716097 -73.997252   160.0   
2                0            6      MANHATTAN  40.733985 -74.005457   233.0   
3                0            7      MANHATTAN  40.716392 -73.983726   163.0   
4                1            9      MANHATTAN  40.726359 -73.988002   273.0   
5               -1           10      MANHATTAN  40.742876 -73.998551   177.0   
6                1           13      MANHATTAN  40.736980 -73.982771   281.0   
7               -1           14      MANHATTAN  40.753830 -73.995050   405.0   
8               -1           17      MANHATTAN  40.756657 -73.970653   178.0   
9                2           18      MANHATTAN  40.765130 -73.985013   302.0   
10               2           19      MANHATTAN  40.767065 -73.963994   370.0   
11              -1           20      MANHATTAN  40.784382 -73.975109   197.0   
12              -1           22      MANHATTAN  40.784413 -73.967138    13.0   
13               2           23      MANHATTAN  40.789068 -73.946818   201.0   
14               3           24      MANHATTAN  40.794691 -73.967040   250.0   
15               1           25      MANHATTAN  40.800868 -73.941000   227.0   
16              -1           26      MANHATTAN  40.814682 -73.956520   118.0   
17               3           28      MANHATTAN  40.808560 -73.952465   193.0   
18              -1           30      MANHATTAN  40.828701 -73.943991   152.0   
19               2           32      MANHATTAN  40.815997 -73.945171   205.0   
20              -1           33      MANHATTAN  40.840992 -73.935974   143.0   
21               2           34      MANHATTAN  40.851259 -73.935131   303.0   
22              -1           40          BRONX  40.810369 -73.925339   419.0   
23               1           41          BRONX  40.816429 -73.895458   183.0   
24              -1           42          BRONX  40.822231 -73.911447   271.0   
25              -1           43          BRONX  40.823195 -73.870354   379.0   
26              -1           44          BRONX  40.837659 -73.920010   390.0   
27               2           45          BRONX  40.830830 -73.827246   189.0   
28              -1           46          BRONX  40.854114 -73.900178   364.0   
29              -1           47          BRONX  40.887270 -73.847293   513.0   
30              -1           48          BRONX  40.844134 -73.900270   331.0   
31               2           49          BRONX  40.855957 -73.844251   258.0   
32               3           50          BRONX  40.883651 -73.902314   214.0   
33              -1           52          BRONX  40.868664 -73.879806   325.0   
34               1           60       BROOKLYN  40.576508 -73.976517   209.0   
35               1           61       BROOKLYN  40.594119 -73.960828   254.0   
36              -1           62       BROOKLYN  40.602419 -74.002898   226.0   
37               3           63       BROOKLYN  40.628032 -73.941419   175.0   
38               0           66       BROOKLYN  40.625653 -73.991293   181.0   
39              -1           67       BROOKLYN  40.648796 -73.950307   473.0   
40               0           68       BROOKLYN  40.639290 -74.022785   187.0   
41               0           69       BROOKLYN  40.648685 -73.905188   179.0   
42              -1           70       BROOKLYN  40.630426 -73.973779   323.0   
43               2           71       BROOKLYN  40.664657 -73.948237   268.0   
44               3           72       BROOKLYN  40.658890 -74.001181   191.0   
45              -1           73       BROOKLYN  40.670889 -73.913899   391.0   
46              -1           75       BROOKLYN  40.671193 -73.881505   704.0   
47              -1           76       BROOKLYN  40.683614 -74.000185   103.0   
48               2           77       BROOKLYN  40.675023 -73.930195   276.0   
49  

In [29]:
create_folium_map()